pike/smb2.py

```diff
@@ -1430,6 +1448,193 @@ def _get_offset(self,owner_sid=None,group_sid=None,sacl=None,dacl=None):
+        acl_count_hole = cur.hole.encode_uint16le(0)
+        # Reserved
+        cur.encode_uint16le(0)
+        for entry in self._entries:
+            entry.encode(cur)
+        if len(self.raw_data):
+            cur.encode_bytes(self.raw_data)
+        if self.acl_size == -1:
+            self.acl_size = cur - self.start
+        acl_size_hole(self.acl_size)
+        if self.ace_count == -1:
+            self.ace_count = len(self._entries)
+        acl_count_hole(self.ace_count)
+
+class NT_ACE(core.Frame):
+
+    def __init__(self, parent=None, end=None):
+        super(NT_ACE, self).__init__(parent)
+        if parent is not None:
+            parent.append(self)
+        self.ace_type = 0
+        self.ace_flags = 0
+        self.ace_size = -1
+        self.access_mask = 0
+        self.sid = None
+        self.raw_data = ''
+        self._entries = []
+
+    def clone(self, new_parent=None):
+        """
+        return a copy of this NT_ACE with calculated fields reset
+        """
+        ace = NT_ACE(new_parent)
+        ace.ace_type = self.ace_type
+        ace.ace_flags = self.ace_flags
+        ace.access_mask = self.access_mask
+        ace.sid = self.sid.clone()
+        if self.raw_data:
+            ace.raw_data = array.array('B', self.raw_data)
+        return ace
+
+    def _decode(self, cur):
+        """
+        decode the ACE
+        """
+        self.ace_type = AceType(cur.decode_uint8le())
+        self.ace_flags = AceFlags(cur.decode_uint8le())
+        self.ace_size = cur.decode_uint16le()
+        self.access_mask = Access(cur.decode_uint32le())
+        self.sid = NT_SID()
+        self.sid.decode(cur)
+        #check if we reach the end
+        ace_remain = self.start + self.ace_size - cur
+        assert ace_remain >= 0
+        if ace_remain:
+            self.raw_data = cur.decode_bytes(ace_remain)
+
+    def _encode(self, cur):
+        """
+        encode the ace
+        """
+        cur.encode_uint8le(self.ace_type)
+        cur.encode_uint8le(self.ace_flags)
+        #hole for ace size
+        ace_size_hole = cur.hole.encode_uint16le(0)
+        cur.encode_uint32le(self.access_mask)
+        self.sid.encode(cur)
+        if len(self.raw_data):
+            cur.encode_bytes(self.raw_data)
+        if self.ace_size == -1:
+            self.ace_size = cur - self.start
+        ace_size_hole(self.ace_size)
+
+class NT_SID(core.Frame):
+
+    def __init__(self):
+        super(NT_SID, self).__init__(self, None)
+        self.raw_data = ''
+        self.revision = 0
+        self.sub_authority_count = -1
+        self.identifier_authority = 0
+        self.sub_authority = []
+
+    def clone(self):
+        """
+        return a copy of this NT_SID with calculated fields reset
+        """
+        sid = NT_SID()
+        sid.revision = self.revision
+        sid.identifier_authority = self.identifier_authority
+        sid.sub_authority = self.sub_authority
```

Comment:
just noticed that this needs to be a copy

```
sid.sub_authority = list(self.sub_authority)
```

or 

```
sid.sub_authority = self.sub_authority[:]
```

Reply:
fixed